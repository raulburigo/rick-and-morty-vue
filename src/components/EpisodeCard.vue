<template>
  <v-lazy
    v-model="isActive"
    :options="{
      threshold: 0.5,
    }"
    min-height="200"
    transition="fade-transition"
  >
    <v-card
      shrink
      elevation="10"
      :class="{
        'text-right': even && !$vuetify.breakpoint.smAndDown,
      }"
    >
      <v-card-title
        primary-title
        class="text-wrap"
        :class="{
          'justify-end': even && !$vuetify.breakpoint.smAndDown,
        }"
        >{{ episode.episode }} - {{ episode.name }}</v-card-title
      >
      <v-card-subtitle
        v-text="
          new Date(Date.parse(episode.air_date)).toLocaleDateString('pt-BR', {
            year: 'numeric',
            month: 'long',
            day: 'numeric',
          })
        "
        :class="{
          'justify-end': even && !$vuetify.breakpoint.smAndDown,
        }"
      ></v-card-subtitle>
      <v-divider class="mx-4"></v-divider>

      <v-card-text>
        <v-row class="pt-1" ref="episode">
          <v-col class="mr-1">
            <v-tooltip
              bottom
              v-for="(char, index) in episode.characters"
              :key="char.id"
            >
              <template v-slot:activator="{ on, attrs }">
                <v-avatar v-show="index <= 5 || show" v-bind="attrs" v-on="on">
                  <v-img :src="char.image"></v-img>
                </v-avatar>
              </template>

              <span>{{ char.name }}</span>
            </v-tooltip>

            <v-btn
              class="my-2 ml-2"
              icon
              outlined
              v-bind:style="{ color: btnColor }"
              v-if="episode.characters.length > 5"
              @click="show = !show"
            >
              <v-icon dark>
                {{ !show ? "mdi-plus" : "mdi-minus" }}
              </v-icon>
            </v-btn>
          </v-col>
        </v-row>
      </v-card-text>
    </v-card>
  </v-lazy>
</template>

<script>
export default {
  name: "EpisodeCard",

  props: { episode: Object, even: Boolean },

  data: function () {
    return {
      show: false,
      isActive: false,
    };
  },

  computed: {
    btnColor: function () {
      return this.$vuetify.theme.dark ? "#80e032" : "#1b661f";
    },
  },
};
</script>
