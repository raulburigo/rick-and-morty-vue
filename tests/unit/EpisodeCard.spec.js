/* eslint-disable no-unused-vars */
import Vue from "vue";
import Vuetify from "vuetify";

import { createLocalVue, mount, shallowMount } from "@vue/test-utils";

import EpisodeCard from "../../src/components/EpisodeCard";

const localVue = createLocalVue();
Vue.use(Vuetify);

describe("EpisodeCard.vue", () => {
  let vuetify;
  const ep1 = {
    episode: "S01E01",
    name: "Pilot",
    air_date: "December 2, 2013",
    characters: [
      {
        id: 1,
        name: "Rick Sanchez",
        image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
      },
      {
        id: 2,
        name: "Morty Smith",
        image: "https://rickandmortyapi.com/api/character/avatar/2.jpeg",
      },
      {
        id: 357,
        name: "Bepisian",
        image: "https://rickandmortyapi.com/api/character/avatar/35.jpeg",
      },
      {
        id: 387,
        name: "Beth Smith",
        image: "https://rickandmortyapi.com/api/character/avatar/38.jpeg",
      },
      {
        id: 627,
        name: "Canklanker Thom",
        image: "https://rickandmortyapi.com/api/character/avatar/62.jpeg",
      },
      {
        id: 927,
        name: "Davin",
        image: "https://rickandmortyapi.com/api/character/avatar/92.jpeg",
      },
      {
        id: 127,
        name: "Frank Palicky",
        image: "https://rickandmortyapi.com/api/character/avatar/127.jpeg",
      },
    ],
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  it("should have a custom title and match snapshot", async () => {
    const wrapper = mount(EpisodeCard, {
      localVue,
      vuetify,
      propsData: { episode: ep1 },
    });

    await wrapper.setData({ isActive: true });
    expect(wrapper.html()).toMatchSnapshot();

    const title = wrapper.find(".v-card__title");

    expect(title.text()).toBe("S01E01 - Pilot");
  });

  it("should render lazyly", async () => {
    const wrapper = mount(EpisodeCard, {
      localVue,
      vuetify,
      propsData: { episode: ep1 },
    });

    const lazyCardBefore = wrapper.find(".v-card");
    expect(lazyCardBefore.exists()).toBe(false);

    await wrapper.setData({ isActive: true });

    const lazyCardAfter = wrapper.find(".v-card");
    expect(lazyCardAfter.exists()).toBe(true);
  });

  it("should render as many avatar as characters and show tooltip with character name ", async (done) => {
    const wrapper = mount(EpisodeCard, {
      localVue,
      vuetify,
      propsData: { episode: ep1 },
    });

    await wrapper.setData({ isActive: true });

    const avatars = wrapper.findAll(".v-avatar");
    expect(avatars.length).toBe(7);

    const rickAvatar = avatars.at(0);
    rickAvatar.trigger("mouseenter");
    await wrapper.vm.$nextTick();
    requestAnimationFrame(() => {
      expect(wrapper.find(".v-tooltip").text()).toEqual("Rick Sanchez");
      done();
    });
  });

  it("should show up to 6 avatars and show/hide more when btn clicked ", async () => {
    const wrapper = mount(EpisodeCard, {
      localVue,
      vuetify,
      propsData: { episode: ep1 },
    });

    await wrapper.setData({ isActive: true });

    const avatars = wrapper.findAll(".v-avatar");
    expect(avatars.length).toBe(7);

    expect(avatars.at(5).isVisible()).toBe(true);

    expect(avatars.at(6).isVisible()).toBe(false);

    const button = wrapper.find("button");
    await button.trigger("click");
    expect(avatars.at(6).isVisible()).toBe(true);

    await button.trigger("click");
    expect(avatars.at(6).isVisible()).toBe(false);
  });
});
