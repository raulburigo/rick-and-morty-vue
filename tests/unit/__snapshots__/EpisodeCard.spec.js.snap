// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeCard.vue should have a custom title and match snapshot 1`] = `
<div class="v-lazy" style="min-height: 200px;">
  <transition-stub name="fade-transition">
    <div class="v-card v-sheet theme--light elevation-10" shrink="">
      <div primary-title="" class="v-card__title text-wrap">S01E01 - Pilot</div>
      <div class="v-card__subtitle">December 2, 2013</div>
      <hr role="separator" aria-orientation="horizontal" class="mx-4 v-divider theme--light">
      <div class="v-card__text">
        <div class="row pt-1">
          <div class="mr-1 col">
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div><span class="v-tooltip v-tooltip--bottom"><!----></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div><span class="v-tooltip v-tooltip--bottom"><!----></span>
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div><span class="v-tooltip v-tooltip--bottom"><!----></span>
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div><span class="v-tooltip v-tooltip--bottom"><!----></span>
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div><span class="v-tooltip v-tooltip--bottom"><!----></span>
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div><span class="v-tooltip v-tooltip--bottom"><!----></span>
            <div class="v-avatar" style="height: 48px; min-width: 48px; width: 48px; display: none;" aria-haspopup="true" aria-expanded="false">
              <div class="v-image v-responsive theme--light">
                <transition-stub name="fade-transition" mode="in-out">
                  <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                </transition-stub>
                <div class="v-responsive__content"></div>
              </div>
            </div> <button type="button" class="my-2 ml-2 v-btn v-btn--icon v-btn--outlined v-btn--round theme--light v-size--default" style="color: rgb(27, 102, 31);"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-plus theme--dark"></i></span></button>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
